### DISCUSSION TITLE
so-rule-update error after update 2.3.130

### DISCUSSION THREAD

**Discussion by madcaban** [Security-Onion-Solutions/securityonion#8250] [Unsupported Versions] [answered]
Hello,

I just upgraded Security Onion to version 2.3.130 and I am having problems launching the "so-rule-update" script.

It returns the following error.

_2022-07-07 12:16:43,706 - <INFO> - Loading ./rulecat.conf.
2022-07-07 12:16:43,709 - <INFO> - Forcing Suricata version to 6.0.
2022-07-07 12:16:43,710 - <INFO> - Fetching https://rules.emergingthreats.net/open/suricata-6.0.0/emerging.rules.tar.gz.
 100% - 3356616/3356616               
2022-07-07 12:16:45,304 - <INFO> - Done.
2022-07-07 12:16:45,502 - <INFO> - Ignoring file rules/emerging-deleted.rules
2022-07-07 12:16:45,503 - <INFO> - Loading local file /opt/so/rules/nids/local.rules
Traceback (most recent call last):
  File "/usr/local/bin/idstools-rulecat", line 12, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.9/site-packages/idstools/scripts/rulecat.py", line 864, in main
    rulemap = build_rule_map(rules)
  File "/usr/local/lib/python3.9/site-packages/idstools/scripts/rulecat.py", line 555, in build_rule_map
    if rule["rev"] > rulemap[rule.id]["rev"]:
TypeError: '>' not supported between instances of 'NoneType' and 'int'__

I have tried to debug the error but I can't find the scripts/folders it mentions (rulecat.py, python3.9, idstools-rulecat...).

Could you please help me with this issue?

Thanks in advance and have a nice day,
David

**Comment by dougburks** [comment]
Have you made any changes to your NIDS rule configuration?

Have you checked for typos?

**Reply by madcaban** [reply]
Hello dougburks,

Thank you very much for your reply. Answering your questions, the only configuration I have done is to add and modify some local rules in the file _/opt/so/saltstack/local/salt/idstools/local.rules_ . The problem has suddenly appeared after upgrading to the new version 2.3.130.

To check for typos, I have commented out all the lines in the local.rules document and the problem still appears (the changes are correctly replicated from _/opt/so/saltstack/local/salt/idstools/local.rules_ to _/opt/so/rules/nids/local.rules_).

I have also detected that there are several files that reference other files that do not exist in that path. In example, in the rulecat.conf file appears the location _/opt/so/idstools/etc/disable.conf_ and this file is located on my machine in _/opt/so/conf/idstools/etc/disable.conf_. The same thing happens with the files that reference the error shown above.

Is there any solution to be able to add local rules in the system?

Thanks for your help and best regards.
David

**Comment by GopalSati** [comment]
Hello,
Iam also having the same issue post upgrading from ver 2.3.110 to ver 2.3.130. After upgrading the SO version I did restore the rules (i backed up the rules before upgrade) under the path /opt/so/saltstack/local/salt/idstools/local.rules.  Please note that my all rules (custom rules i have created) were properly working before upgrading the SO. Please advise the solution.

[root@siem01 ~]# sudo so-rule-update
2022-07-08 15:21:21,886 - <INFO> - Loading ./rulecat.conf.
2022-07-08 15:21:21,889 - <INFO> - Forcing Suricata version to 6.0.
2022-07-08 15:21:21,890 - <INFO> - Fetching https://rules.emergingthreats.net/open/suricata-6.0.0/emerging.rules.tar.gz.
 100% - 3373897/3373897
2022-07-08 15:21:22,222 - <INFO> - Done.
2022-07-08 15:21:22,436 - <INFO> - Ignoring file rules/emerging-deleted.rules
2022-07-08 15:21:22,437 - <INFO> - Loading local file /opt/so/rules/nids/local.rules
2022-07-08 15:21:22,437 - <INFO> - Loading local file /opt/so/rules/nids/sorules/extraction.rules
2022-07-08 15:21:22,437 - <INFO> - Loading local file /opt/so/rules/nids/sorules/filters.rules
2022-07-08 15:21:25,707 - <ERROR> - failed to parse rule:
Traceback (most recent call last):
  File "/usr/local/bin/idstools-rulecat", line 12, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.9/site-packages/idstools/scripts/rulecat.py", line 861, in main
    rules += idstools.rule.parse_fileobj(
  File "/usr/local/lib/python3.9/site-packages/idstools/rule.py", line 361, in parse_fileobj
    rule = parse(buf + line, group)
  File "/usr/local/lib/python3.9/site-packages/idstools/rule.py", line 317, in parse
    rule[name] = int(val)
ValueError: invalid literal for int() with base 10: '3 classtype:tcp-event'
[root@siem01 ~]#

**Reply by dougburks** [reply]
@GopalSati From https://github.com/Security-Onion-Solutions/securityonion/discussions/1720:
> Start a new discussion instead of replying to somebody else's discussion
Please search to see if you can find similar discussions that may help you. However, in order to avoid confusion, please do NOT reply to somebody else's discussion with your own issue. Instead, please start a new discussion and in that new discussion you can provide a hyperlink to the related discussion.

**Comment by xfaith** [comment, marked as answer]
The paths you are referencing are inside the docker container for so-idstools and are binded  
 "Binds": [
                "/opt/so/conf/idstools/etc:/opt/so/idstools/etc:ro",
                "/opt/so/rules/nids:/opt/so/rules/nids:rw"

You could access it with docker exec -it so-idstools /bin/bash  - but I would not recommend this.  I think its something that occured on the host machine, something custom.


  Not 100% sure of this but it sounds like its might be a duplicate SID (I have not been able to replicate this though using local.rules at the least.  Maybe something to do with modify rules.  

--From you statement you are modifying rules in local.rules?  Not using modifysid?  

But additional information from suricata-update was based on idstools, looks like it has a bit more error checking.  From that file

if rule["rev"] > rulemap[rule.id]["rev"]:
                 logger.warning(
                    "Found duplicate rule SID {}, "
                   "keeping the rule with greater revision.".format(rule.sid))
                   rulemap[rule.id] = rule

[GopalSati] - I would look at your rules also, do you have local rules?  Look for a classtype that does not have a ; before it.  Looks like its trying to parse a number along with classtype

**Reply by madcaban** [reply]
Hello,

You mean that there is a local.rules that has a SID the same as another rule? In my case I have commented out and deleted all the rules in the file and the problem is still being experienced.

In my case, I am only modifying the _/opt/so/saltstack/local/salt/idstools/local.rules_ file.

Regarding your suggestion, I have checked the existing rules and they all have a "**;**" after the "classtype" field, so it does not appear that this is the problem (it also happens when the local.rules file is empty).

If you need some more information to check this problem, I am at your disposal.

Best regards and thanks

**Reply by dougburks** [reply]
As a test, you could try backing up your `local.rules` file and then reverting it back to default configuration:
```
# Custom Suricata rules go in this file
```
(and nothing else)

**Reply by madcaban** [reply]
Hi dougburks,

I created a new _local.rules_ file as you suggested and was able to launch the _so-rule-update_ command.

After that, I pasted the local rules into this file and was able to relaunch the command without any problem.

I assumed there was something wrong with the old file.

Thanks for your help and have a nice day,
David
